src/polygon.c

```diff
@@ -554,6 +557,58 @@ spoly_segment(SLine *sl, const SPOLY *poly, int32 i)
 	}
 }
 
+static bool
+poly_get_point(SPoint *sp, const SPOLY *poly, int32 i)
+{
+	if (i >= 0 && i < poly->npts)
+	{
+		memcpy((void *) sp, (void *) &poly->p[i], sizeof(SPoint));
+		return true;
+	}
+	return false;
+}
+
+Datum
+spherepoly_get_point(PG_FUNCTION_ARGS)
+{
+	int32		i;
+	SPOLY		*poly = PG_GETARG_SPOLY(0);
+	SPoint		*sp = (SPoint *) palloc(sizeof(SPoint));
+
+	i = PG_GETARG_INT32(1);
+	if (poly_get_point(sp, poly, i - 1))
+	{
+		PG_RETURN_POINTER(sp);
+	}
+	pfree(sp);
+	PG_RETURN_NULL();
+}
+
+Datum
+spherepoly_get_array(PG_FUNCTION_ARGS)
+{
+    SPOLY *poly = PG_GETARG_SPOLY(0);
+    Datum *datum_arr = (Datum *) palloc(sizeof(Datum) * poly->npts);
+    ArrayType *res;
+	SPoint *p = (SPoint *) palloc(sizeof(SPoint) * poly->npts);
+    for (size_t i = 0; i < poly->npts; i++)
+    {
+        if (!poly_get_point(&p[i], poly, i))
+        {
+            // Clean up and return NULL
+            for (size_t j = 0; j < i; j++)
+                pfree(DatumGetPointer(datum_arr[j]));
+            pfree(datum_arr);
+            PG_RETURN_NULL();
+        }
+        datum_arr[i] = PointerGetDatum(&p[i]);
+    }
+
+    res = construct_array(datum_arr, poly->npts, TypenameGetTypid("spoint"), sizeof(SPoint), false, 'd');
```

Comment:
TypenameGetTypid should not be called every time when spherepoly_get_array is called because it is time consuming operation. The idea is to cache spoint type in the process memory (in a static variable) id and reuse it.